[GStreamer] Properly report video frame pixel format

philn opened the pull request:
#### fb297855a9fc6b37f8d1ec2faaa09f6eaa365d2a
<pre>
[GStreamer] Properly report video frame pixel format
<a href="https://bugs.webkit.org/show_bug.cgi?id=248953">https://bugs.webkit.org/show_bug.cgi?id=248953</a>

Reviewed by Xabier Rodriguez-Calvar.

The VideoFrameGStreamer::pixelFormat() method is now properly implemented and the mapping between
GStreamer and WebCore pixel formats is fulfilled.

* Source/WebCore/platform/VideoPixelFormat.cpp:
(WebCore::convertVideoFramePixelFormat):
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.h:

Canonical link: <a href="https://commits.webkit.org/257909@main">https://commits.webkit.org/257909@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c06c5612b9afbf3d615671985edd9d1b1761cec2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100382 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9544 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33445 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109698 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169937 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104375 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10456 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/96 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92786 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107576 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106160 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91180 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/34566 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/22571 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77523 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3285 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/24090 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3277 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/212 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9398 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43580 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5432 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5093 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/VideoPixelFormat.cpp:
```diff
@@ -43,6 +47,34 @@ VideoPixelFormat convertVideoFramePixelFormat(uint32_t format, bool shouldDiscar
+    case GST_VIDEO_FORMAT_BGRA:
+        return shouldDiscardAlpha ? VideoPixelFormat::BGRX : VideoPixelFormat::BGRA;
+    case GST_VIDEO_FORMAT_ARGB:
+        return shouldDiscardAlpha ? VideoPixelFormat::RGBX : VideoPixelFormat::RGBA;
```
Is this correct or a copy&paste error?

philn replied:
looks fishy indeed

philn replied:
Ah I think I did the same as on the Cocoa code path... But I suspect we'd instead need a formal `ARGB` in `VideoPixelFormat`. WDYT @youennf ?

youennf replied:
`VideoPixelFormat` is bound to WebIDL defined https://www.w3.org/TR/webcodecs/#enumdef-videopixelformat, which does not have ARGB.
What we have done in Cocoa is to relate WebCodecs RGBA to CVPixelBuffer ARGB, see Cocoa VideoFrame::createRGBA.

philn replied:
I suppose we'll have to do the same then :)

calvaris commented on Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.cpp:
```diff
@@ -271,6 +272,17 @@ void VideoFrame::paintInContext(GraphicsContext& context, const FloatRect& desti
+uint32_t VideoFrameGStreamer::pixelFormat() const
+{
+    if (m_cachedVideoFormat != GST_VIDEO_FORMAT_UNKNOWN)
+        return m_cachedVideoFormat;
```
I wonder if the video format could change if the caps change, for example, rendering this cache incorrect at that point. Maybe invalidating the cache at some other point?

philn replied:
a video frame wraps a sample basically, so I don't see in which case the sample caps would change.

ceyusa commented on Source/WebCore/platform/VideoPixelFormat.cpp:
```diff
@@ -26,6 +26,10 @@
 #include "VideoPixelFormat.h"
 
+#if USE(GSTREAMER)
+#include <gst/video/video-format.h>
```
afaik, the supported include is `gst/video/video.h`

philn replied:
video-format.h is public too, and has smaller impact :)

ceyusa commented on Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.h:
```diff
@@ -23,6 +23,7 @@
 
 #include "VideoFrame.h"
 #include "VideoFrameMetadataGStreamer.h"
+#include <gst/video/video-format.h>
```
ditto

philn committed:
[GStreamer] Properly report video frame pixel format
https://bugs.webkit.org/show_bug.cgi?id=248953

Reviewed by Xabier Rodriguez-Calvar.

The VideoFrameGStreamer::pixelFormat() method is now properly implemented and the mapping between
GStreamer and WebCore pixel formats is fulfilled.

* Source/WebCore/platform/VideoPixelFormat.cpp:
(WebCore::convertVideoFramePixelFormat):
* Source/WebCore/platform/graphics/gstreamer/VideoFrameGStreamer.h:

Canonical link: https://commits.webkit.org/257909@main
